chore(api): apply mypy consistently across api

aaron-kulkarni opened the pull request:
<!--
Thanks for taking the time to open a Pull Request (PR)! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

GitHub provides robust markdown to format your PR. Links, diagrams, pictures, and videos along with text formatting make it possible to create a rich and informative PR. For more information on GitHub markdown, see:

https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview

At present, typechecking is not applied fully and consistently across the API directory. This PR removes exceptions to mypy checking rules and makes sure that all mypy typechecking tests pass.

EXEC-201 EXEC-202

<!--
Describe your PR at a high level. State acceptance criteria and how this PR fits into other work. Link issues, PRs, and other relevant resources.
-->

## Test Plan and Hands on Testing

<!--
Describe your testing of the PR. Emphasize testing not reflected in the code. Attach protocols, logs, screenshots and any other assets that support your testing.
-->

Remove all exceptions from `api/mypy.ini` and run `make -C api lint` from the home directory and make sure all the tests pass.

## Changelog

<!--
List changes introduced by this PR considering future developers and the end user. Give careful thought and clear documentation to breaking changes.
-->

## Review requests

<!--
- What do you need from reviewers to feel confident this PR is ready to merge?
- Ask questions.
-->

Lots of file changes in this one. Most changes in most files are just monotonous `-> None` function return type annotations(or something similar). For reviewers, I would focus on the following files:


`protocols/labware.py`: My edits to _check_for_subdirectories seem strange. I'm not confident in the new things I imported to get that type annotation to work.

`test_transfers.py`: There are dozens of conversions between `Location`, `Well`, and `Location | Well`. What should be done about these? If one function has a `Well` and wants to call a function that takes in a `Location`, how should I convert it?

## Risk assessment

<!--
- Indicate the level of attention this PR needs.
- Provide context to guide reviewers.
- Discuss trade-offs, coupling, and side effects.
- Look for the possibility, even if you think it's small, that your change may affect some other part of the system.
  - For instance, changing return tip behavior may also change the behavior of labware calibration.
- How do your unit tests and on hands on testing mitigate this PR's risks and the risk of future regressions?
- Especially in high risk PRs, explain how you know your testing is enough.
-->

Low.

aaron-kulkarni committed:
chore(api): apply mypy consistently across api

At present, typechecking is not applied fully and consistently across the API directory. This PR
removes exceptions to mypy checking rules and makes sure that all mypy typechecking tests pass.

EXEC-201 EXEC-202

aaron-kulkarni committed:
protocol_api_old done

aaron-kulkarni committed:
opentrons/system done

aaron-kulkarni committed:
failing test fixes and tempdeck changes

aaron-kulkarni committed:
random file fixes. only 598 issues remaining!

aaron-kulkarni committed:
more random file fixes. 546 left.

aaron-kulkarni committed:
Merge branch 'edge' into exec-201-strict-mypy-api

aaron-kulkarni committed:
format fixes, other file changes. ~500 errors to go

aaron-kulkarni committed:
~450 left

aaron-kulkarni committed:
eod changes, minor

aaron-kulkarni committed:
multiple test file fixes. <400 left

aaron-kulkarni committed:
execute_json fixes. 275 remaining

aaron-kulkarni committed:
test_instruments file. ~215 errors left

aaron-kulkarni committed:
~125 left

aaron-kulkarni committed:
mostly done except for estimator. 112 left

aaron-kulkarni committed:
estimator done. 35 left

SyntaxColoring commented on api/tests/opentrons/system/test_wifi.py:
```diff
@@ -159,7 +159,7 @@ def test_eap_check_option():
 
 
-async def test_list_keys(wifi_keys_tempdir):
+async def test_list_keys(wifi_keys_tempdir: str) -> None:
```
```suggestion
async def test_list_keys(wifi_keys_tempdir: Path) -> None:
```

SyntaxColoring commented on api/tests/opentrons/system/test_wifi.py:
```diff
@@ -177,7 +177,7 @@ async def test_list_keys(wifi_keys_tempdir):
 
 
-async def test_key_lifecycle(wifi_keys_tempdir):
+async def test_key_lifecycle(wifi_keys_tempdir: str) -> None:
```
```suggestion
async def test_key_lifecycle(wifi_keys_tempdir: Path) -> None:
```

aaron-kulkarni committed:
just transfers.py still has errors. only 10 left

aaron-kulkarni committed:
remove pdb and mypy.ini comments

sfoster1 commented on api/src/opentrons/protocols/labware.py:
```diff
@@ -19,6 +19,9 @@
 from opentrons_shared_data.labware.types import LabwareDefinition
 
+if TYPE_CHECKING:
+    from _typeshed import GenericPath
```
this does seem weird. Can it be `pathlib.Path`?

aaron-kulkarni replied:
I tried a few other ways: using path, using string, etc. but couldn't seem to get everything to fit together properly.

sfoster1 reviewed (commented):
Going from `Well` to `Location` is done by doing `well.top()` or `well.bottom()` or something. In general though, I wouldn't worry too much about it; just getting the warnings removed is good enough, and making the typing elegant can happen in a followup pass.

The things to worry about are actual non-typechecking code changes to production software. That's what needs a second look, as in the API transfers code, for instance.

aaron-kulkarni committed:
various fixes/changes. no errors.

aaron-kulkarni committed:
all tests pass?

aaron-kulkarni committed:
test fix

aaron-kulkarni committed:
format

ryanthecoder commented on api/tests/opentrons/hardware_control/test_instruments.py:
```diff
@@ -164,43 +199,47 @@ async def test_mismatch_fails(sim_and_instr):
     hw_api_cntrlr = await API.build_hardware_controller(
         loop=asyncio.get_running_loop(),
         feature_flags=HardwareFeatureFlags.build_from_ff(),
+        firmware=_find_smoothie_file(),
```
Should Be NONE since this isn't running on an actual robot

aaron-kulkarni replied:
Talked about in person - made a different change

aaron-kulkarni committed:
done

sfoster1 committed:
no code changes for max_many

sfoster1 committed:
no internals for get_all_labware_definitions

sfoster1 committed:
protocols/labware: remove TYPE_CHECKING

sfoster1 committed:
protocols/labware: no code changes

sfoster1 committed:
tests/.../test_types: remove commented code

sfoster1 committed:
protocols/execution/dev_types: correct asdict

sfoster1 committed:
protocols/api_support/util: fix axis dict

This was a bit more in depth - the code changes here actually needed a
bit more, as much as I wanted to remove them.

sfoster1 committed:
protocols/advanced_control/transfers: dedupe type

sfoster1 committed:
protocols/advanced_control/transfers: correct iter

sfoster1 committed:
transfers: comments and noise

sfoster1 reviewed (approved):
Reviewed this, fixed up some little changes and some noise, very happy with it

SyntaxColoring commented on api/src/opentrons/protocols/advanced_control/transfers.py:
```diff
@@ -414,6 +436,8 @@ def __init__(
         # list of Wells into a 1 dimensional list of Wells
         pipette_configuration_type = NozzleConfigurationType.FULL
+        sources: List[Union[Well, types.Location]]
+        dests: List[Union[Well, types.Location]]
```
Nitpick, optional: if `sources` and `dests` are distinct from `srcs` and `dsts`, could we have more distinct names for them? `normalized_srcs` and `normalized_dsts` maybe?

SyntaxColoring commented on api/src/opentrons/protocols/api_support/util.py:
```diff
@@ -397,8 +393,11 @@ def _check_version_wrapper(*args: Any, **kwargs: Any) -> Any:
-    def __contains__(self, item):
+class ModifiedList(list[str]):
+    def __contains__(self, item: object) -> bool:
+        # This weird thing where item is passed in as an object
+        # then asserts that it's a string is a workaround for MyPy errors.
+        assert isinstance(item, str)
```
Nitpick, optional:

I think this mypy error that we're "working around" is actually a legitimate typing bug. Calling code can do the expression `123 in ModifiedList(["a", "b", "c"])`, and one would normally expect that to just return `False`, but this code raises an `AttributeError` before this PR or an `AssertionError` with this PR.

I guess we should have it return `False`?

sfoster1 replied:
Yeah, agreed. We should. It's a problem with how the magic dunders are specified in the language.

SyntaxColoring commented on api/src/opentrons/protocols/execution/execute_python.py:
```diff
@@ -57,7 +58,9 @@ def _find_protocol_error(tb, proto_name):
 
 def _raise_pretty_protocol_error(exception: Exception, filename: str) -> None:
-    exc_type, exc_value, tb = sys.exc_info()
+    _, _, tb = sys.exc_info()
+    if tb is None:
+        raise Exception
```
Is this meant to be `raise exception` (lowercase `e`), so it's handled like the `except KeyError` case a few lines below?

sfoster1 replied:
yes, I think so

sfoster1 replied:
i think a better way to handle it is for `_find_protocol_error` to allow `None`

SyntaxColoring commented on api/src/opentrons/protocols/labware.py:
```diff
@@ -98,6 +98,7 @@ def save_definition(
     :param location: File path
     """
     namespace = labware_def["namespace"]
+    assert "loadName" in labware_def["namespace"]
```
Is this `assert` correct? This looks like it will assert that `"loadName"` is a substring of whatever is the string value for `namespace` (typically `"opentrons"` or `"custom_beta"`)?

sfoster1 replied:
yeah wait what how did I miss this. This should just get deleted

SyntaxColoring reviewed (changes requested):
This mostly looks good to me, but here's a few things that look questionable.

sfoster1 committed:
test_wifi: path instead of str

sfoster1 committed:
transfers: better naming

sfoster1 committed:
execute_python: fix exceptions

sfoster1 committed:
protocols/labware: bad assert

sfoster1 committed:
protocols/api_support/util: better in

SyntaxColoring reviewed (approved):
Hallelujah.

sfoster1 commented:
The failing tests are because fe6252c268f51658e5de6e017cf6985e1f1f0e85 needs to be merged into edge as well as release

sfoster1 committed:
Merge branch 'edge' into exec-201-strict-mypy-api
